src/snowflake/connector/cache.py

```diff
@@ -507,19 +527,46 @@ def _load(self) -> bool:
         """Load cache from disk if possible, returns whether it was able to load."""
         try:
             with open(self.file_path, "rb") as r_file:
-                other = pickle.load(r_file)
-            self._update(
-                other,
+                other: SFDictFileCache = pickle.load(r_file)
+            # Since we want to know whether we are dirty after loading
+            #  we have to know whether the file could learn anything from self
+            #  so instead of calling self.update we call other.update and swap
+            #  the 2 underlying caches after.
+            self._lock.release()
+            cache_file_learnt = other.update(
+                self,
                 update_newer_only=True,
             )
+            self._lock.acquire()
```

Comment:
I can understand that the lock release then reacquire here is because other.update will lock the lock.

I forget why we didn't choose to use a re-entrant lock? those lock operations make me a bit concerned of deadlock. I wonder if RLock could simplify the lock logic here

Reply:
`RLock` is slower than `Lock` is all.